File: src/main/java/no/echokarriere/backend/user/UserEntity.java
package no.echokarriere.backend.user;

import lombok.Builder;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import no.echokarriere.graphql.types.CreateUserInput;
import no.echokarriere.graphql.types.UpdateUserInput;

import java.time.OffsetDateTime;
import java.util.UUID;

@Value
@Builder
@RequiredArgsConstructor
public class UserEntity {
    UUID id;
    String name;
    String email;
    String password;
    boolean active;
    UserType userType;
    OffsetDateTime createdAt;
    OffsetDateTime modifietAt;

    public UserEntity(CreateUserInput input) {
        this.id = UUID.randomUUID();
        this.name = input.getName();
        this.email = input.getEmail();
        // TODO: create random initial password
        this.password = "";
        this.active = true;
        this.userType = convertUserType(input.getUserType());
        this.createdAt = OffsetDateTime.now();
        this.modifietAt = null;
    }

    public UserEntity(UUID id, UpdateUserInput input) {
        this.id = id;
        this.name = input.getName();
        this.email = input.getEmail();
        // TODO: Skip this field when updating?
        this.password = "";
        this.active = true;
        this.userType = convertUserType(input.getUserType());
        this.createdAt = null;
        this.modifietAt = OffsetDateTime.now();
    }

    private UserType convertUserType(no.echokarriere.graphql.types.UserType userType) {
        if (userType == no.echokarriere.graphql.types.UserType.ADMIN) return UserType.ADMIN;
        else if (userType == no.echokarriere.graphql.types.UserType.STAFF) return UserType.STAFF;
        else return UserType.USER;
    }
}

File: src/main/kotlin/no/echokarriere/user/UserEntity.kt
package no.echokarriere.user

import no.echokarriere.enums.Usertype
import org.jooq.impl.EnumConverter
import java.time.OffsetDateTime
import java.util.UUID

enum class UserType {
    ADMIN,
    STAFF,
    USER
}

@Suppress("unused") // used by Jooq
class UserTypeConverter : EnumConverter<UserType, Usertype>(UserType::class.java, Usertype::class.java)

data class User(private val entity: UserEntity) {
    val id = entity.id
    val name = entity.name
    val email = entity.email
    val active = entity.active
    val type = entity.type
}

class UserEntity private constructor(
    val id: UUID = UUID.randomUUID(),
    val name: String,
    val email: String,
    val password: String,
    val active: Boolean,
    val type: UserType,
    val createdAt: OffsetDateTime,
    val modifiedAt: OffsetDateTime? = null
) {
    companion object {
        fun create(
            id: UUID = UUID.randomUUID(),
            name: String,
            email: String,
            password: String,
            active: Boolean,
            type: UserType,
            createdAt: OffsetDateTime = OffsetDateTime.now(),
            modifiedAt: OffsetDateTime? = null
        ): UserEntity = UserEntity(
            id = id,
            name = name,
            email = email,
            password = password,
            active = active,
            type = type,
            createdAt = createdAt,
            modifiedAt = modifiedAt
        )
    }
}

data class CreateUserInput(
    val name: String,
    val email: String,
    val password: String,
    val type: UserType = UserType.USER
) {
    fun createEntity(): UserEntity = UserEntity.create(
        name = name,
        email = email,
        password = password,
        type = type,
        active = true
    )
}
